src/Browser.ts

```diff
@@ -0,0 +1,312 @@
+     const targetPromise = new Promise<Puppeteer.Target>(x => resolve = x);
+     this.on(Events.Browser.TargetCreated, check);
+     this.on(Events.Browser.TargetChanged, check);
+     try {
+       if (!timeout)
+         return await targetPromise;
+       return await helper.waitWithTimeout<Puppeteer.Target>(targetPromise, 'target', timeout);
+     } finally {
+       this.removeListener(Events.Browser.TargetCreated, check);
+       this.removeListener(Events.Browser.TargetChanged, check);
+     }
+
+     function check(target: Puppeteer.Target): void  {
+       if (predicate(target))
+         resolve(target);
+     }
+   }
+
+   /**
+   * @return {!Promise<!Array<!Puppeteer.Page>>}
+   */
+   async pages(): Promise<Puppeteer.Page[]> {
+     const contextPages = await Promise.all(this.browserContexts().map(context => context.pages()));
+     // Flatten array.
+     return contextPages.reduce((acc, x) => acc.concat(x), []);
+   }
+
+   async version(): Promise<string> {
+     const version = await this._getVersion();
+     return version.product;
+   }
+
+   async userAgent(): Promise<string> {
+     const version = await this._getVersion();
+     return version.userAgent;
+   }
+
+   async close(): Promise<void> {
+     await this._closeCallback.call(null);
+     this.disconnect();
+   }
+
+   disconnect(): void {
+     this._connection.dispose();
+   }
+
+   isConnected(): boolean {
+     return !this._connection._closed;
+   }
+
+   _getVersion(): Promise<Protocol.Browser.getVersionReturnValue> {
+     return this._connection.send('Browser.getVersion');
+   }
+}
+
+export class BrowserContext extends EventEmitter {
+  _connection: Connection;
+  _browser: Browser;
+  _id?: string;
+
+  constructor(connection: Connection, browser: Browser, contextId?: string) {
+    super();
+    this._connection = connection;
+    this._browser = browser;
+    this._id = contextId;
+  }
+
+  targets(): Puppeteer.Target[] {
+    return this._browser.targets().filter(target => target.browserContext() === this);
+  }
+
+  waitForTarget(predicate: (x: Puppeteer.Target) => boolean, options: { timeout?: number}): Promise<Puppeteer.Target> {
+    return this._browser.waitForTarget(target => target.browserContext() === this && predicate(target), options);
+  }
+
+  async pages(): Promise<Puppeteer.Page[]> {
+    const pages = await Promise.all(
+        this.targets()
+            .filter(target => target.type() === 'page')
+            .map(target => target.page())
+    );
+    return pages.filter(page => !!page);
+  }
+
+  isIncognito(): boolean {
+    return !!this._id;
+  }
+
+  async overridePermissions(origin: string, permissions: Protocol.Browser.PermissionType[]): Promise<void> {
+    const webPermissionToProtocol = new Map<string, Protocol.Browser.PermissionType>([
+      ['geolocation', 'geolocation'],
+      ['midi', 'midi'],
+      ['notifications', 'notifications'],
+      // TODO: push isn't a valid type?
+      // ['push', 'push'],
```

Comment:
@mathiasbynens this is interesting to me. The protocol doesn't list `push` as a valid permission type so I'm not quite sure if:

* the protocol is wrong
* Puppeteer is wrong and calling overridePermissions with `push` was effectively a no-op.

I think I'm lacking context here so I can't really make a call. We can keep the code and loosen the types to `string[]` rather than `PermissionType[]` but if this is highlighting a bug it's worth exploring.

Reply:
I lack the context here as well. Let's stick with the existing behavior for now, and we can resolve this separately later. Thanks for adding the TODO.